compiler/noirc_evaluator/src/ssa/opt/mem2reg.rs

```diff
@@ -599,10 +647,32 @@ impl<'f> PerFunctionContext<'f> {
                             if let Some(aliases) = references.aliases.get_mut(expression) {
                                 // The argument reference is possibly aliased by this block parameter
                                 aliases.insert(*parameter);
+
+                                // Check if we have seen the same argument
+                                if let Some(seen_parameters) = arg_set.get_mut(&argument) {
+                                    // Add the current parameter to the parameter we have seen for this argument.
+                                    // The previous parameters and the current one alias one another.
+                                    seen_parameters.insert(*parameter);
+                                } else {
+                                    let mut new_param_set = HashSet::default();
+                                    new_param_set.insert(*parameter);
+                                    arg_set.insert(argument, new_param_set);
+                                }
                             }
                         }
                     }
                 }
+
+                // Set the aliases of the parameters
+                for (_, aliased_params) in arg_set {
+                    for param in aliased_params.iter() {
+                        self.set_aliases(
```

Comment:
We were adding parameters as aliases to the arguments, but we were missing actually whether the parameters alias each other if they have been passed the same argument.